### PR TITLE
Resolve fully qualified storefront URLs

### PR DESCRIPTION
**Pull request by makyo**
This allows passing a storefront URL (eg: https://jujucharms.com/u/charmers/mysql/trusty/28 - see url_test.go for permutations) to parseReference.  This is the first step to supporting v2 charm/bundle URLs

### CODE REVIEW

File: url.go [hunk 1]
```diff
@@ -152,64 +157,106 @@ func ParseReference(url string) (*Reference, error) {
...
-		if r.Schema != "cs" && r.Schema != "local" {
-			return nil, fmt.Errorf("entity URL has invalid schema: %q", url)
+	// Check if we're dealing with a v1 or v2 URL.
+	// TODO: this is a naive check that only works on fully-qualified URLs;
+	// further work will be required for future URLs
+	// XXX Get rid of urlVersion, test based on validStore.MatchString
+	if validStore.MatchString(url) {
+		r.Schema = "cs"
+		url = validStore.ReplaceAllString(url, "")
+		parts := strings.Split(url, "/")
```
Comment (frankban) [inline url.go]: What about something like `parts := strings.Split(strings.Trim(url), "/")` so that we don't need to strip leading/trailing slashes later?

**Comment by frankban** [comment]
:+1: with the change in tests as discussed.
Thanks!

File: url.go [hunk 2]
```diff
@@ -152,64 +157,111 @@ func ParseReference(url string) (*Reference, error) {
...
-	i := strings.Index(url, ":")
-	if i >= 0 {
-		r.Schema = url[:i]
-		if r.Schema != "cs" && r.Schema != "local" {
-			return nil, fmt.Errorf("entity URL has invalid schema: %q", url)
+	// Check if we're dealing with a v1 or v2 URL.
+	// TODO: this is a naive check that only works on fully-qualified URLs;
+	// further work will be required for future URLs
+	// XXX Get rid of urlVersion, test based on validStore.MatchString
+	if validStore.MatchString(url) {
```
Comment (rogpeppe) [inline url.go]: Perhaps split both the arms of this if statement into
two functions (parseNewStyleReference and parseOldStyleReference?).

I'd call parseNewStyleReference iff the url starts with https?: - then
the error messages will be sensible if the url is a new style URL
but not entirely well formed.

File: url.go [hunk 3]
```diff
@@ -152,64 +157,111 @@ func ParseReference(url string) (*Reference, error) {
...
+			if len(parts) < 3 {
+				return nil, fmt.Errorf("entity URL malformed, expecting user and name: %q", url)
+			}
+			r.User = parts[1]
+			parts = parts[2:]
+		}
+		r.Name = parts[0]
+		parts = parts[1:]
+		r.Revision = -1
+		if len(parts) > 0 {
```
Comment (rogpeppe) [inline url.go]: If this was a separate function, you could do:

   if len(parts) == 0 {
       return &r, nil
    }

File: url.go [hunk 4]
```diff
@@ -152,64 +157,111 @@ func ParseReference(url string) (*Reference, error) {
...
+				if !IsValidSeries(r.Series) {
+					return nil, fmt.Errorf("entity URL has invalid series: %q", url)
+				}
+				parts = parts[1:]
+				if len(parts) == 1 {
+					r.Revision, err = strconv.Atoi(parts[0])
+					if err != nil {
+						return nil, fmt.Errorf("entity URL has malformed revision: %q in %q", parts[0], url)
+					}
+				} else {
```
Comment (rogpeppe) [inline url.go]: } else if {

File: url.go [hunk 5]
```diff
@@ -49,6 +53,7 @@ var ErrUnresolvedUrl error = fmt.Errorf("entity url series is not resolved")
 var (
 	validSeries = regexp.MustCompile("^[a-z]+([a-z0-9]+)?$")
 	validName   = regexp.MustCompile("^[a-z][a-z0-9]*(-[a-z0-9]*[a-z][a-z0-9]*)*$")
+	validStore  = regexp.MustCompile(fmt.Sprintf("^https?://(www\\.)?%s/", charmstoreURL))
```
Comment (rogpeppe) [inline url.go]: I'd be tempted use url.Parse to parse
the URL, and then do the rest of the checking on the parts
of the parsed URL.

File: url.go [hunk 6]
```diff
@@ -151,18 +152,35 @@ func ParseReference(url string) (*Reference, error) {
 }
 
 func parseReference(url string) (*Reference, error) {
+	// Check if we're dealing with a v1 or v2 URL.
+	u, err := gourl.Parse(url)
+	if err != nil {
+		return nil, fmt.Errorf("entity URL is not a valid url: %s", url)
```
Comment (rogpeppe) [inline url.go]: perhaps:

    return fmt.Errorf("cannot parse entity URL %q: %v", url, err)

?

File: url.go [hunk 7]
```diff
@@ -151,18 +152,35 @@ func ParseReference(url string) (*Reference, error) {
...
+	}
+	// Shortcut old-style URLs.
+	if u.Opaque != "" {
```
Comment (rogpeppe) [inline url.go]: Cool, I always saw Opaque in passing and didn't know quite what it was. Now I know! Very useful for our purposes.

File: url.go [hunk 8]
```diff
@@ -151,18 +152,35 @@ func ParseReference(url string) (*Reference, error) {
 }
 
 func parseReference(url string) (*Reference, error) {
+	// Check if we're dealing with a v1 or v2 URL.
+	u, err := gourl.Parse(url)
+	if err != nil {
+		return nil, fmt.Errorf("entity URL is not a valid url: %s", url)
+	}
```
Comment (rogpeppe) [inline url.go]: Perhaps:

     if u.FawQuery != "" || u.Fragment != "" || u.User != nil {
          return fmt.Errorf("entity URL %s has unrecognised parts", url)
     }

?

File: url.go [hunk 9]
```diff
@@ -211,6 +226,61 @@ func parseReference(url string) (*Reference, error) {
...
+	return &r, nil
+}
+
+func parseV2Reference(url *gourl.URL) (*Reference, error) {
+	var r Reference
+	r.Schema = "cs"
+	parts := strings.Split(strings.Trim(url.Path, "/"), "/")
+	if len(parts) == 0 {
+		return nil, fmt.Errorf("entity URL malformed, no parts: %q", url)
```
Comment (rogpeppe) [inline url.go]: This can never happen (strings.Split always returns at least one element in the slice)

File: url.go [hunk 10]
```diff
@@ -211,6 +226,61 @@ func parseReference(url string) (*Reference, error) {
...
+	}
+	if parts[0] == "u" {
+		if len(parts) < 3 {
+			return nil, fmt.Errorf("entity URL malformed, expecting user and name: %q", url)
```
Comment (rogpeppe) [inline url.go]: return fmt.Errorf("malformed entity URL %q; expected \"/u/<user>/<name>\"", url)

?

File: url.go [hunk 11]
```diff
@@ -211,6 +226,61 @@ func parseReference(url string) (*Reference, error) {
...
+		}
+		r.User = parts[1]
```
Comment (rogpeppe) [inline url.go]: an idiom that can look nice:

    r.User, parts = parts[1], parts[2:]

(and below)
YMMV.

File: url.go [hunk 12]
```diff
@@ -151,18 +152,35 @@ func ParseReference(url string) (*Reference, error) {
 }
 
 func parseReference(url string) (*Reference, error) {
```
Comment (rogpeppe) [inline url.go]: Please update the doc comment on ParseReference to include the new forms of URL.

**Comment by rogpeppe** [comment]
LGTM thanks!

**Comment by makyo** [comment]
Thanks all

**Comment by makyo** [comment]
$$merge$$

**Pull request merged**